src/api_manager/config_manager.cc

```diff
@@ -20,86 +20,71 @@ namespace google {
-static std::vector<std::pair<std::string, int>> kEmptyConfigs;
 }  // namespace
 
 ConfigManager::ConfigManager(
     std::shared_ptr<context::GlobalContext> global_context,
     RolloutApplyFunction rollout_apply_function)
     : global_context_(global_context),
       rollout_apply_function_(rollout_apply_function),
-      refresh_interval_ms_(kCheckNewRolloutInterval) {
+      fetch_throttle_window_in_s_(kFetchThrottleWindowInS) {
   if (global_context_->server_config() &&
       global_context_->server_config()->has_service_management_config()) {
-    // update refresh interval in ms
+    // update fetch_throttle_window
     if (global_context_->server_config()
             ->service_management_config()
-            .refresh_interval_ms() > 0) {
-      refresh_interval_ms_ = global_context_->server_config()
-                                 ->service_management_config()
-                                 .refresh_interval_ms();
+            .fetch_throttle_window_s() > 0) {
+      fetch_throttle_window_in_s_ = global_context_->server_config()
+                                        ->service_management_config()
+                                        .fetch_throttle_window_s();
     }
   }
+  // throttle in milliseconds
+  random_dist_.reset(new std::uniform_int_distribution<int>(
+      0, fetch_throttle_window_in_s_ * 1000));
 
   service_management_fetch_.reset(new ServiceManagementFetch(global_context));
 }
 
 ConfigManager::~ConfigManager() {
-  if (rollouts_refresh_timer_) {
-    rollouts_refresh_timer_->Stop();
+  if (fetch_timer_) {
+    fetch_timer_->Stop();
   }
 };
 
-void ConfigManager::Init() {
-  if (global_context_->rollout_strategy() == kRolloutStrategyManaged &&
-      refresh_interval_ms_ > 0) {
-    // Counter window is 5 time of refresh_interval. It means
-    // only there is not any requests in the last 5 intervals,
-    // stop checking rollout.
-    window_request_counter_.reset(new utils::TimeBasedCounter(
-        kWindowCounterSize, std::chrono::milliseconds(refresh_interval_ms_ * 5),
-        std::chrono::system_clock::now()));
-    rollouts_refresh_timer_ = global_context_->env()->StartPeriodicTimer(
-        std::chrono::milliseconds(refresh_interval_ms_),
-        [this]() { OnRolloutsRefreshTimer(); });
+void ConfigManager::SetLatestRolloutId(
+    const std::string& latest_rollout_id,
+    std::chrono::time_point<std::chrono::system_clock> now) {
+  if (latest_rollout_id == current_rollout_id_) {
+    return;
   }
-}
 
-void ConfigManager::OnRolloutsRefreshTimer() {
-  global_context_->env()->LogDebug("RolloutsRefreshTimer start");
-  // If there is not any requests in the last window, not to call.
-  if (window_request_counter_ &&
-      window_request_counter_->Count(std::chrono::system_clock::now()) == 0) {
-    global_context_->env()->LogDebug(
-        "RolloutsRefresh skipped due to no traffic.");
+  // Timer is pending, or too close to last fetch time
+  if (fetch_timer_ || now < next_window_time_) {
```

Comment:
so logic decides whether to fetch new rollout or not, right?

should it be 

        if (!fetch_timer_ || now < next_window_time_) ?

Reply:
No,  the opposite, do nothing.  If condition meets, do nothing, return.   if fetch_timer is still pending,  do nothing.  We only need one fetch timer.   when timer is fired,  fetch_timer will be null,  but it still have some time to fetch the service config.  so we set up next_window_time = timer_fired_time + window_size.    So we will not fire another timer before next window.